### PR TITLE
Itsi interactives in learn portal

### PR DESCRIPTION
**Pull request by sagartalla**
Adding ITSI models in the Learn portal.

### CODE REVIEW

**Comment by scytacki** [comment]
@sagartalla can you update this, so it is mergable?
@pjanik can you review this? It is a very large set of changes unfortunately.

File: features/teacher_searches_groups_and_previews_instructional_materials.feature [hunk 1]
```diff
@@ -156,32 +156,6 @@ Feature: Teacher can search instructional materials grouped by material type, so
...
-    And the Activity "intersecting lines" is assigned to the class "Geography"
-    And the Activity "intersecting lines" is assigned to the class "Mathematics"
-    And the Activity "parallel lines" is assigned to the class "Mathematics"
-    And the Activity "parallel lines" is assigned to the class "Geography"
-    And I follow "Popularity" in Sort By on the search instructional materials page
-    And I should wait 2 seconds
-    Then "intersecting lines" should appear before "parallel lines"
-
-
-  @javascript
```
Comment (pjanik) [inline features/teacher_searches_groups_and_previews_instructional_materials.feature]: Why are these tests removed?

File: app/models/interactive.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+   'Interactive'
+  end
+
+  def teacher_only
+    true
+  end
+  
+  def teacher_only?
+    true
+  end
```
Comment (pjanik) [inline app/models/interactive.rb]: Do we need all these methods (starting from `is_template`)? If they are only used by SOLR search, I would just use `searchable` block above.

Also, do we need all properties in `searchable` block (e.g. ones that return `nil`, like `teacher_description`)?

Not a big deal, but I would keep interface minimal so it's clear what's actually needed.

File: app/helpers/search_helper.rb [hunk 3]
```diff
@@ -38,9 +48,9 @@ def build_onSearch_message(form_model)
   end
 
   def show_material_icon(material, link_url, hide_details)
-    if material.material_type == "Investigation"
-      icon_url = material.icon_image
-    elsif material.material_type == "Activity"
+    if material.material_type == "Interactive"
+      icon_url = material.image_url
```
Comment (pjanik) [inline app/helpers/search_helper.rb]: A tiny thing, but I would just rename `image_url` or provide `icon_image` alias.

File: app/assets/javascripts/import_model_library.js.erb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+var lightboxconfig={
+        content:"<div style='padding:10px'>Loading...Please Wait.</div>",
+        title:"Import Model Library",
+        width:400,
+        height:160
+    },import_model_library_modal;
+
+function ajax_success(data){
+   import_model_library_modal.handle.setContent(data);
+};
```
Comment (pjanik) [inline app/assets/javascripts/import_model_library.js.erb]: I would wrap this whole file in self-executing function, so we don't expose too many unnecessary globals. `ajax_success` can be risky, especially if this code is based on some existing one. :wink: Also, small thing, but I would keep CamelCase naming, typical whitespace (` ` before `{` etc.) and fix semicolons.

**Comment by pjanik** [comment]
In general looks good to me, it follows existing patterns. I've left a few inline comments regarding some details.

Maybe the biggest issue is that there are no specs. Perhaps at least basic ones would be useful.

And a few more details:
- When I create an interactive with URL without any protocol (no `http(s)://`), `Run interactive` link doesn't work.
- When there is no image, interactive doesn't look nice on the index page. :wink:
- Can I get to `/interactives/import_model_library` using UI? But maybe it wasn't necessary.
- Also, I couldn't import interactives that I exported a moment before (JS error: `Uncaught ReferenceError: import_model_library_post_json is not defined`). Should it work in such case or does it work only with interactives exported from ITSI?

**Comment by scytacki** [comment]
The export from the portal should also import in the portal. That should fixed if it is broken.

**Comment by sagartalla** [comment]
@pjanik The import of interactives from Portal can be done by clicking the Import Interactives button present in the Authoring section of the portal.(You need to login as an Admin)Are you trying to import interactives from here?
Today again we have tested Importing an interactive to a portal that it was exported from and its working fine.

@scytacki We can handle the broken image issue by removing the image tag for interactives that don't have image source.
Also,interactives without an http(s) protocol work with relative url's(//path).Should we handle this issue in Interactive authoring itself ,where,if a user enters a url without the protocols, we will make such a url relative.Is that okay?

File: features/teacher_searches_groups_and_previews_instructional_materials.feature (hunk 1 above)
Comment (sagartalla) [inline features/teacher_searches_groups_and_previews_instructional_materials.feature]: @pjanik We have a story for removing the popularity filter here https://www.pivotaltracker.com/story/show/91504150.

**Comment by scytacki** [comment]
- Yes if there is no image please remove the image tag, but make sure things still line up vertically.
- We don't need to explicitly support relative URLS, but I'm curious why they don't work. Are you checking for the `http` protocol at the beginning somewhere?  

In general we will encourage authors to use https URLs. But the current ITSI image resources are at an http only site, and that is OK for now.

**Comment by pjanik** [comment]
> We don't need to explicitly support relative URLS, but I'm curious why they don't work. Are you checking for the http protocol at the beginning somewhere?

I was lazy and I typed `lab.concord.org` - I guess it was treated as a relative url. Not a big deal probably, but adding `//` or `http://` by default can prevent some authors from being confused.

**Comment by scytacki** [comment]
@sagartalla, this looks good to me. Can you merge master into it again and resolve any conflicts?

@psndcsrv, this will probably affect your search results because it is adding a new Interactives category.

**Comment by sagartalla** [comment]
@scytacki Here are the screenshots of the filter area in the browse material page
with_project_filter with_project_filter
with_interactive_filter with_interactive_filter

We are thinking of moving the authorship filter (from the second screenshot) under the project filter (in the first screenshot). Is it fine?

**Comment by scytacki** [comment]
Putting the authorship in the same column as the Project filter sounds good.
And you'll have Subject Areas and Grade Levels in the same column right?

**Comment by jengoree** [comment]
There's this comment in the project filtering story:
*We already have 3 projects in the Learn portal (Genigames, Geniverse,
InquirySpace, possibly also NextGen MW), so this list could get long in the
next couple of years. On the search page we could move Grade Levels under
Sort By, and use the Grade Level Column for Projects. Either that or we
will have to use a pull-down like we are right now for Interactive Types
(seehttp://has.staging.concord.org/search
<http://has.staging.concord.org/search>), and put that under Authorship.*

I think what Sagar is suggesting is fine, but we're going to have to keep
an eye on that project list as it's potentially going to grow quickly. (If
we published all the activities we have now this summer to Learn, we'll
have Genigames, Geniverse, InquirySpace, NextGen MW, Interactions, RITES,
HAS, ITSI.) That will pretty much take up all the current vertical space
already.
We should also keep in mind that we're going to have at least 6 lines for
subject areas, plus potentially 5 for grade level.

Jen Goree
jgoree@concord.org
Software Portfolio and Project Manager
The Concord Consortium

**Comment by sagartalla** [comment]
@scytacki Do we need to have project assignment for interactives?

Filter area after merge looks like this screenshot from 2015-05-07 18 17 18

**Comment by scytacki** [comment]
yes please add project assignment, and merge in master again to resolve conflicts.

**Comment by scytacki** [comment]
I went ahead and merged master in. If it builds cleanly and looks ok on my machine I'll merge and deploy deploy learn.staging.
In the last week the search page was refactored to use react.js, this is not in master yet. So that code will need to be updated to support this interactive work.

**Pull request merged**